apps/full-site-editing/full-site-editing-plugin/donations/edit.js

```diff
@@ -11,34 +11,64 @@ import Donations from './donations';
 import LoadingError from './loading-error';
 import LoadingStatus from './loading-status';
 import UpgradePlan from './upgrade-plan';
+import fetchDefaultProducts from './fetch-default-products';
 import fetchStatus from './fetch-status';
 
 const Edit = () => {
 	const [ isLoading, setIsLoading ] = useState( true );
 	const [ loadingError, setLoadingError ] = useState( '' );
 	const [ shouldUpgrade, setShouldUpgrade ] = useState( false );
 	const [ stripeConnectUrl, setStripeConnectUrl ] = useState( false );
+	const [ products, setProducts ] = useState( [] );
 	const [ upgradeUrl, setUpgradeUrl ] = useState( '' );
 
+	const apiError = ( message ) => {
+		setLoadingError( message );
+		setIsLoading( false );
+	};
+
+	const hasRequiredProducts = ( resultProducts ) => {
+		const USDDonations = resultProducts.filter(
+			( product ) => product.currency === 'USD' && product.type === 'donation'
+		);
+
+		if ( ! USDDonations.length ) {
+			return;
+		}
+
+		const intervals = [];
+		USDDonations.forEach( ( donation ) => intervals.push( donation.interval ) );
```

Comment:
Since we want a map+filter, [Array.reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce) is probably more suitable. This is mostly a micro-optimization to avoid a second pass on the array so take it as optional. It'd look something like: 

```jsx
const hasDonationProducts = ( products, currency ) => {
    const productIntervals = products.reduce( ( ( intervals, { currency: productCurrency, type, interval } ) => {
        if ( currency === productCurrency  && type === 'donation' ) {
              intervals.push( interval );
        }
        return intervals;
    } ), [] );
    //... then return true or false with includes or a set difference.
}
```

Reply:
Yup, agree. Pushed out a commit using reduce for building the array, but might've been lost during [last outage](https://www.githubstatus.com/incidents/d60fmldlvw48). Let's see if it's back after the recovery

Reply:
Ah, thank you both! I've never really grokked `reduce` 🙃